docs/source/contribution/threading.rst

```diff
@@ -0,0 +1,166 @@
+The last argument is the lambda function that defines a function object that proceeds ``i``-th
+iteration of the loop.
+
+Blocking
+--------
+
+To have more control over the parallel execution and to increase
+`cache locality <https://en.wikipedia.org/wiki/Locality_of_reference>`_ oneDAL usually splits
+the data into blocks and then processes those blocks in parallel.
+
+This code shows how a typical parallel loop in oneDAL looks like:
+
+.. include:: ../includes/threading/sum-parallel-by-blocks.rst
+
+Thread-local Storage (TLS)
+**************************
+
+Lets consider you need to compute a dot product of two arrays.
+Here is a variant of sequential implementation:
+
+.. include:: ../includes/threading/dot-sequential.rst
+
+Parallel computations can be performed in two steps:
+
+    1. Compute partial dot product at each threaded.
+    2. Perform a reduction: Add the partial results from all threads to compute the final dot product.
+
+``daal::tls`` provides a local storage where each thread can accumulate its local results.
+Following code allocates memory that would store partial dot products for each thread:
+
+.. include:: ../includes/threading/dot-parallel-init-tls.rst
+
+``SafeStatus`` in this code denotes a thread-safe counterpart of the ``Status`` class.
+``SafeStatus`` allows to collect errors from all threads and report them to user using
+``detach()`` method as it will be shown later in the code.
+
+Checking the status right after the initialization code won't show the allocation errors,
+because oneTBB uses lazy evaluation and the lambda function passed to the constructor of the TLS
+is evaluated in the moment of the TLS's first use.
+
+Again, there are several options available in the threading layer of oneDAL to compute the partial
+dot product results at each thread.
+One of the options is to use the already mentioned ``daal::threader_for`` and blocking approach
+as shown here:
+
+.. include:: ../includes/threading/dot-parallel-partial-compute.rst
+
+To compute the final result it is requred to reduce TLS's partial results over all threads
+as shown here:
+
+.. include:: ../includes/threading/dot-parallel-reduction.rst
+
+Local memory of the threads should also be released when it is no longer needed.
+
+The complete parallel verision of dot product computations would look like:
+
+.. include:: ../includes/threading/dot-parallel.rst
+
+Static Work Scheduling
+**********************
+
+By default, oneTBB uses
+`dynamic work scheduling <https://oneapi-src.github.io/oneTBB/main/tbb_userguide/How_Task_Scheduler_Works.html>`_
+and work stealing.
+It means that two different runs of the same parallel loop can produce different
+mappings of the loop's iteration space to the available threads.
+This strategy is beneficial when it is difficult to estimate the amount of work performed
+by each iteration.
+
+In the cases when it is known that the iterations perform an equal amount of work, it
```

Comment:
In all cases, where you know how much work there is to do in the parallel regions, it will be faster to map the work onto threads using a dynamic schedule. Drop the use of `it might be beneficial to use ...` in favour of `it is more performant to use ...`